root/account/applications/Register.js

```diff
@@ -0,0 +1,93 @@
+/*
+ * @flow
+ * Copyright (C) 2018 MetaBrainz Foundation
+ *
+ * This file is part of MusicBrainz, the open internet music database,
+ * and is licensed under the GPL version 2, or (at your option) any
+ * later version: http://www.gnu.org/licenses/gpl-2.0.txt
+ */
+
+import * as React from 'react';
+import noop from 'lodash/noop';
+
+import FormRow from '../../components/FormRow';
+import FormRowSelect from '../../components/FormRowSelect';
+import FormRowText from '../../components/FormRowText';
+import FormRowURLLong from '../../components/FormRowURLLong';
+import FormSubmit from '../../components/FormSubmit';
+import Layout from '../../layout';
+import {l, N_l} from '../../static/scripts/common/i18n';
+
+type OauthTypeT = 'installed' | 'web';
+
+type ApplicationRegisterForm = FormT<{|
+  +name: FieldT<string>,
+  +oauth_redirect_uri: FieldT<string>,
+  +oauth_type: FieldT<OauthTypeT>,
+|}>;
+
+type Props = {|
+  +form: ApplicationRegisterForm,
+|};
+
+type State = {|
+  +withURL: boolean,
+|};
+
+const oauthTypeOptions = {
+  grouped: false,
+  options: [
+    {label: N_l('Web Application'), value: 'web'},
+    {label: N_l('Installed Application'), value: 'installed'},
+  ],
+};
+
+class RegisterApplication extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+    this.state = {withURL: (props.form.field.oauth_type.value || oauthTypeOptions.options[0].value) === 'web'};
+    //this.handleOauthTypeChange = this.handleOauthTypeChange.bind(this);
+  }
+
+  //handleOauthTypeChange: (event: SyntheticEvent<HTMLSelectElement>) => void;
+
+  handleOauthTypeChange(e: SyntheticEvent<HTMLSelectElement>) {
+    (this: any).setState({withURL: e.currentTarget.value === 'web'});
+  }
+
+  render() {
+    return (
+      <Layout fullWidth title={l('Register Application')}>
+        <div id="content">
+          <h1>{l('Register Application')}</h1>
+          <form action="/account/applications/register" method="post">
+            <FormRowText
+              field={this.props.form.field.name}
+              label={l('Name:')}
+              required
+            />
+            <FormRowSelect
+              field={this.props.form.field.oauth_type}
+              label={l('Type:')}
+              onChange={this.handleOauthTypeChange}
```

Comment:
That is supposed to update the component state and thus to show/hide the URL field.

Reply:
Let me know if root/utility/hydrate.js makes sense. Basically you'll need to factor out a component just to hold the `<form />` in a separate file, say `RegisterApplicationForm`, since the rest of the page is static. Then `export default hydrate('register-application', RegisterApplicationForm);` from there. The first argument `'register-application'` is the class name it'll use for the wrapper div that acts as the mount point for `RegisterApplicationForm`.